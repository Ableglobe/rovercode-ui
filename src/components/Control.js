import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import { Button } from 'semantic-ui-react';
import { hot } from 'react-hot-loader';
import PropTypes from 'prop-types';

import {
  changeExecutionState as actionChangeExecutionState,
  EXECUTION_RUN,
  EXECUTION_STEP,
  EXECUTION_STOP,
  EXECUTION_RESET,
} from '@/actions/code';

const mapStateToProps = ({ code }) => ({ code });
const mapDispatchToProps = dispatch => ({
  changeExecutionState: state => dispatch(actionChangeExecutionState(state)),
});

const Control = ({ code, changeExecutionState }) => (
  <Fragment>
    {
      code.execution === EXECUTION_RUN ? (
        <Button onClick={() => changeExecutionState(EXECUTION_STOP)}>
          Stop
        </Button>
      ) : (
        <Fragment>
          <Button onClick={() => changeExecutionState(EXECUTION_STEP)}>
            Step
          </Button>
          <Button onClick={() => changeExecutionState(EXECUTION_RUN)}>
            Run
          </Button>
          <Button onClick={() => changeExecutionState(EXECUTION_RESET)}>
            Reset
          </Button>
        </Fragment>
      )
    }
  </Fragment>
);

Control.propTypes = {
  code: PropTypes.shape({
    execution: PropTypes.number,
  }).isRequired,
  changeExecutionState: PropTypes.func.isRequired,
};

export default hot(module)(connect(mapStateToProps, mapDispatchToProps)(Control));
