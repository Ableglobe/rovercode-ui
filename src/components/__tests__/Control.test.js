import React from 'react';
import { Button } from 'semantic-ui-react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';

import {
  changeExecutionState,
  EXECUTION_RUN,
  EXECUTION_STEP,
  EXECUTION_STOP,
  EXECUTION_RESET,
} from '@/actions/code';
import Control from '../Control';

describe('The Control component', () => {
  const mockStore = configureStore();
  let store;

  beforeEach(() => {
    store = mockStore({
      code: {
        execution: EXECUTION_STOP,
      },
    });
    store.dispatch = jest.fn();
  });

  test('renders on the page with no errors', () => {
    const wrapper = mount(<Control store={store} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('dispatches an action to change execution state', () => {
    const wrapper = shallow(<Control store={store} />);

    wrapper.props().changeExecutionState(EXECUTION_RUN);

    expect(store.dispatch).toHaveBeenCalledWith(changeExecutionState(EXECUTION_RUN));
  });

  test('changes to step state on button press', () => {
    const wrapper = shallow(<Control store={store} />);

    wrapper.dive().find(Button).at(0).simulate('click');

    expect(store.dispatch).toHaveBeenCalledWith(changeExecutionState(EXECUTION_STEP));
  });

  test('changes to run state on button press', () => {
    const wrapper = shallow(<Control store={store} />);

    wrapper.dive().find(Button).at(1).simulate('click');

    expect(store.dispatch).toHaveBeenCalledWith(changeExecutionState(EXECUTION_RUN));
  });

  test('changes to reset state on button press', () => {
    const wrapper = shallow(<Control store={store} />);

    wrapper.dive().find(Button).at(2).simulate('click');

    expect(store.dispatch).toHaveBeenCalledWith(changeExecutionState(EXECUTION_RESET));
  });

  test('changes to stop state on button press', () => {
    const runStore = mockStore({
      code: {
        execution: EXECUTION_RUN,
      },
    });
    runStore.dispatch = jest.fn();
    const wrapper = shallow(<Control store={runStore} />).dive();

    wrapper.find(Button).at(0).simulate('click');

    expect(runStore.dispatch).toHaveBeenCalledWith(changeExecutionState(EXECUTION_STOP));
  });
});
