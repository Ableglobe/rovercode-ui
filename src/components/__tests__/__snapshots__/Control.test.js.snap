// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Control component renders on the page with no errors 1`] = `
<Connect(Control)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Control
    changeExecutionState={[Function]}
    code={
      Object {
        "execution": 3,
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [MockFunction],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  >
    <Button
      as="button"
      onClick={[Function]}
      role="button"
    >
      <button
        className="ui button"
        onClick={[Function]}
        role="button"
      >
        Step
      </button>
    </Button>
    <Button
      as="button"
      onClick={[Function]}
      role="button"
    >
      <button
        className="ui button"
        onClick={[Function]}
        role="button"
      >
        Run
      </button>
    </Button>
    <Button
      as="button"
      onClick={[Function]}
      role="button"
    >
      <button
        className="ui button"
        onClick={[Function]}
        role="button"
      >
        Reset
      </button>
    </Button>
  </Control>
</Connect(Control)>
`;
